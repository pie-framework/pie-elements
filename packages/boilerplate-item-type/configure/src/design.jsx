import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import cloneDeep from 'lodash/cloneDeep';
import { InputContainer, settings, layout } from '@pie-lib/pie-toolbox/config-ui';
import {EditableHtml} from '@pie-lib/pie-toolbox/editable-html';
import {test, callTest} from 'pie-utils';

callTest('configure');

const { Panel, toggle } = settings;

export class Design extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    configuration: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired,
    onModelChanged: PropTypes.func.isRequired,
    onConfigurationChanged: PropTypes.func.isRequired,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired,
    }),
    uploadSoundSupport: PropTypes.object,
  };

  static defaultProps = {};

  onPromptChanged = (prompt) => {
    const { onModelChanged, model } = this.props;
    const update = cloneDeep(model);

    update.prompt = prompt;
    onModelChanged(update);
  };

  render() {
    const { classes, configuration, imageSupport, model, onConfigurationChanged, onModelChanged, uploadSoundSupport } =
      this.props;
    const {
      contentDimensions = {},
      prompt = {},
      settingsPanelDisabled,
    } = configuration || {};
    const { promptEnabled } = model || {};

    const panelSettings = {};

    const panelProperties = {
      promptEnabled: prompt.settings && toggle(prompt.label),
    };

    return (
      <layout.ConfigLayout
        dimensions={contentDimensions}
        hideSettings={settingsPanelDisabled}
        settings={
          <Panel
            model={model}
            configuration={configuration}
            onChangeModel={(model) => onModelChanged(model)}
            onChangeConfiguration={onConfigurationChanged}
            groups={{
              Settings: panelSettings,
              Properties: panelProperties,
            }}
          />
        }
      >
        {test}
        {promptEnabled && (
          <InputContainer label={prompt.label || ''} className={classes.promptHolder}>
            <EditableHtml
              className={classes.prompt}
              markup={model.prompt}
              onChange={this.onPromptChanged}
              imageSupport={imageSupport}
              uploadSoundSupport={uploadSoundSupport}
              languageCharactersProps={[{ language: 'spanish' }, { language: 'special' }]}
            />
          </InputContainer>
        )}
      </layout.ConfigLayout>
    );
  }
}

export default withStyles((theme) => ({
  promptHolder: {
    width: '100%',
    paddingTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2,
  }
}))(Design);
