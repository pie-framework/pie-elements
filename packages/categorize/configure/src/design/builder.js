import cloneDeep from 'lodash/cloneDeep';

export const buildCategories = (categories, choices, correctResponse) => {
  const clonedCategories = cloneDeep(categories);

  return clonedCategories.map(category => {
    const cr = correctResponse.find(cr => cr.category === category.id);

    if (cr) {
      category.choices = (cr.choices || []).map(choiceId => {
        const choice = choices.find(h => h.id === choiceId);
        if (choice) {
          return Object.assign({}, { id: choice.id, content: choice.content });
        }
      });
    }

    return category;
  });
};

const getChoices = (cat, choices, index) => {
  if (!cat.alternateResponses || cat.alternateResponses.length === 0) {
    return [];
  }

  return (cat.alternateResponses[index] || []).map(alt => {
    return choices.find(ch => ch.id === alt);
  });
};

export const buildAlternateResponses = (categories, choices, correctResponse) => {
  const mostAlternates = correctResponse.reduce((mostAlt, cat) => {
    if (cat.alternateResponses && cat.alternateResponses.length >= mostAlt) {
      return cat.alternateResponses.length;
    }

    return mostAlt;
  }, 0);
  const alternatesArray = (new Array(mostAlternates)).fill(0);

  return alternatesArray.map((val, index) => {
    return correctResponse.map((cat, catIndex) => {
      const currentCategory = categories[catIndex];

      return {
        id: currentCategory.id,
        label: currentCategory.label,
        choices: getChoices(cat, choices, index)
      };
    });
  });
};
