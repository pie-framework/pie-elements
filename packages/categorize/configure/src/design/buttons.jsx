import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Button from '@material-ui/core/Button';
import MuiDivider from '@material-ui/core/Divider';

export class RawAddButton extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    label: PropTypes.string,
    onClick: PropTypes.func,
    disabled: PropTypes.bool
  };

  static defaultProps = {
    label: 'Add'
  };

  render() {
    const { classes, className, label, onClick, disabled } = this.props;
    return (
      <Button
        onClick={onClick}
        disabled={disabled}
        className={classNames(classes.addButton, className)}
        size="small"
        variant="contained"
        color="primary"
      >
        {label}
      </Button>
    );
  }
}
const styles = theme => ({
  addButton: {
    height: theme.spacing.unit * 4
  }
});

const AddButton = withStyles(styles)(RawAddButton);

const DeleteButton = withStyles(theme => ({
  deleteButton: {
    margin: 0,
    padding: 0
  }
}))(({ classes, label, onClick }) => (
  <Button
    className={classes.deleteButton}
    onClick={onClick}
    size="small"
    color="primary"
  >
    {label}
  </Button>
));

const Divider = withStyles(theme => ({
  divider: {
    marginTop: 2 * theme.spacing.unit,
    marginBottom: 2 * theme.spacing.unit
  }
}))(({ classes }) => <MuiDivider className={classes.divider} />);

export { AddButton, DeleteButton, Divider };
