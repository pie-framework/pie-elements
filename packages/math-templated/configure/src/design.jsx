import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import cloneDeep from 'lodash/cloneDeep';
import pick from 'lodash/pick';
import throttle from 'lodash/throttle';
import {
  InputContainer,
  settings,
  layout,
} from '@pie-lib/pie-toolbox/config-ui';
import {
  ALL_PLUGINS,
  EditableHtml,
} from '@pie-lib/pie-toolbox/editable-html';
import { dropdown } from '@pie-lib/pie-toolbox/code/config-ui/settings';

import Response from './response';
import { processMarkup } from './markupUtils';

const { Panel, toggle } = settings;

const createElementFromHTML = (htmlString) => {
  const div = document.createElement('div');
  div.innerHTML = htmlString.trim();

  return div;
};

class Design extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    configuration: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired,
    onModelChanged: PropTypes.func.isRequired,
    onConfigurationChanged: PropTypes.func.isRequired,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired,
    }),
    uploadSoundSupport: PropTypes.object,
  };

  state = {};

  componentDidMount() {
    const { model: { slateMarkup } } = this.props;

    this.setState({ markup: slateMarkup });

  }

  handleChange = (key, value) => {
    const { onModelChanged, model } = this.props;
    const updatedModel = cloneDeep(model);
    updatedModel[key] = value;
    onModelChanged(updatedModel);
  };

  onResponseChange = (response, index) => {
    const { model, onModelChanged } = this.props;
    const newModel = { ...model };

    newModel.responses[index] = response;
    onModelChanged(newModel);
  };

  onChangeResponse = (index, newVal) => {
    const { model, onModelChanged } = this.props;
    const { responses } = model;

    if (!responses[index]) {
      responses[index] = [{ answer: newVal || '', id: 'response' + index, allowSpaces: true }];
    } else {
      responses[index][0].answer = newVal || '';
    }

    onModelChanged({ ...model, responses });
  };

  onResponsesChanged = (responses) => {
    this.props.onModelChanged({ ...this.props.model, responses });
  };

  onChange = (markup) => {
    const {
      model: { responses },
      onModelChanged,
    } = this.props;
    const newResponses = {};
    const domMarkup = createElementFromHTML(markup);
    const responseAreas = domMarkup.querySelectorAll('[data-type="math_templated"]');

    responseAreas.forEach((element) => {
      const { value, index: dataIndex } = element.dataset;

      if (!value) {
        element.dataset.value = `R ${dataIndex}`;
      }

      newResponses[dataIndex] = responses[dataIndex] || {
        // TODO: Initialize with customizable defaults
        allowSpaces: true,
        validation: 'symbolic',
        allowTrailingZeros: false,
        ignoreOrder: false,
        answer: '',
        alternates: {},
      };

      element.dataset.index = dataIndex;
    });

    const processedMarkup = processMarkup(markup);

    const callback = () =>
      onModelChanged({
        ...this.props.model,
        slateMarkup: domMarkup.innerHTML,
        responses: newResponses,
        markup: processedMarkup
      });

    this.setState({ cachedChoices: undefined }, callback);
  };

  onHandleAreaChange = throttle(
    (nodes) => {
      const {
        model: { responses },
        onModelChanged,
      } = this.props;
      const { cachedResponses } = this.state;

      if (!nodes) {
        return;
      }

      const newChoices = responses ? cloneDeep(responses) : {};
      const newCachedResponses = cachedResponses ? cloneDeep(cachedResponses) : {};

      nodes.forEach((node) => {
        const keyForNode = node.data.get('index');

        if (!newChoices[keyForNode] && newCachedResponses[keyForNode]) {
          Object.assign(newChoices, pick(newCachedResponses, keyForNode));

          if (newCachedResponses.hasOwnProperty(keyForNode)) {
            delete newCachedResponses[keyForNode];
          }
        } else {
          Object.assign(newCachedResponses, pick(newChoices, keyForNode));

          if (newChoices.hasOwnProperty(keyForNode)) {
            delete newChoices[keyForNode];
          }
        }
      });

      const callback = () => onModelChanged({ ...this.props.model, responses: newChoices });

      this.setState({ cachedResponses: newCachedResponses }, callback);
    },
    500,
    { trailing: false, leading: true },
  );

  onBlur = (e) => {
    const { relatedTarget, currentTarget } = e || {};

    function getParentWithRoleTooltip(element, depth = 0) {
      // only run this max 16 times
      if (!element || depth >= 16) return null;

      const parent = element.offsetParent;

      if (!parent) return null;

      if (parent.getAttribute('role') === 'tooltip') {
        return parent;
      }

      return getParentWithRoleTooltip(parent, depth + 1);
    }

    function getDeepChildDataKeypad(element, depth = 0) {
      // only run this max 4 times
      if (!element || depth >= 4) return null;

      const child = element?.children?.[0];

      if (!child) return null;

      if (child.attributes && child.attributes['data-keypad']) {
        return child;
      }

      return getDeepChildDataKeypad(child, depth + 1);
    }

    const parentWithTooltipRole = getParentWithRoleTooltip(relatedTarget);
    const childWithDataKeypad = parentWithTooltipRole ? getDeepChildDataKeypad(parentWithTooltipRole) : null;

    if (!relatedTarget || !currentTarget || !childWithDataKeypad?.attributes['data-keypad']) {
      this.setState({ activeAnswerBlock: '' });
    }
  };

  render() {
    const {
      classes,
      configuration,
      imageSupport,
      model,
      onConfigurationChanged,
      onModelChanged,
      uploadSoundSupport,
    } = this.props;

    const {
      baseInputConfiguration = {},
      contentDimensions = {},
      prompt = {},
      rationale = {},
      settingsPanelDisabled,
      teacherInstructions = {},
      language = {},
      languageChoices = {},
      spellCheck = {},
      playerSpellCheck = {},
      maxImageWidth = {},
      maxImageHeight = {},
      mathMlOptions = {},
      template = {},
      editSource = {},
      ignoreOrder: cIgnoreOrder = {},
      allowTrailingZeros: cAllowTrailingZeros = {},
      partialScoring = {},
      maxResponseAreas,
    } = configuration || {};

    const {
      errors,
      promptEnabled,
      rationaleEnabled,
      spellCheckEnabled,
      teacherInstructionsEnabled,
      toolbarEditorPosition,
      responses,
      equationEditor,
    } = model || {};

    const {
      choices: choicesErrors = {},
      prompt: promptError,
      rationale: rationaleError,
      responseAreas: responseAreasError,
      teacherInstructions: teacherInstructionsError,
      responses: responsesErrors,
    } = errors || {};

    const panelSettings = {
      'language.enabled': language.settings && toggle(language.label, true),
      language: language.settings && language.enabled && dropdown(languageChoices.label, languageChoices.options),
    };

    const panelProperties = {
      teacherInstructionsEnabled: teacherInstructions.settings && toggle(teacherInstructions.label),
      rationaleEnabled: rationale.settings && toggle(rationale.label),
      promptEnabled: prompt.settings && toggle(prompt.label),
      spellCheckEnabled: spellCheck.settings && toggle(spellCheck.label),
      playerSpellCheckEnabled: playerSpellCheck.settings && toggle(playerSpellCheck.label),
      'editSource.enabled': editSource?.settings && toggle(editSource.label, true),
      partialScoring: partialScoring.settings && toggle(partialScoring.label),
    };

    const defaultImageMaxWidth = maxImageWidth && maxImageWidth.prompt;
    const defaultImageMaxHeight = maxImageHeight && maxImageHeight.prompt;

    const toolbarOpts = {
      position: toolbarEditorPosition === 'top' ? 'top' : 'bottom',
    };

    const getPluginProps = (props = {}, baseInputConfiguration = {}) => ({
      ...baseInputConfiguration,
      ...props,
    });

    return (
      <layout.ConfigLayout
        dimensions={contentDimensions}
        hideSettings={settingsPanelDisabled}
        settings={
          <Panel
            model={model}
            configuration={configuration}
            onChangeModel={(updatedModel) => onModelChanged(updatedModel)}
            onChangeConfiguration={onConfigurationChanged}
            groups={{
              Settings: panelSettings,
              Properties: panelProperties,
            }}
          />
        }
      >
        {teacherInstructionsEnabled && (
          <InputContainer
            label={teacherInstructions.label}
            className={classes.promptHolder}
          >
            <EditableHtml
              className={classes.prompt}
              markup={model.teacherInstructions || ''}
              onChange={(value) => this.handleChange('teacherInstructions', value)}
              imageSupport={imageSupport}
              nonEmpty={false}
              error={teacherInstructionsError}
              toolbarOpts={toolbarOpts}
              pluginProps={getPluginProps(
                teacherInstructions?.inputConfiguration,
                baseInputConfiguration
              )}
              spellCheck={spellCheckEnabled}
              maxImageWidth={
                (maxImageWidth && maxImageWidth.teacherInstructions) ||
                defaultImageMaxWidth
              }
              maxImageHeight={
                (maxImageHeight && maxImageHeight.teacherInstructions) ||
                defaultImageMaxHeight
              }
              uploadSoundSupport={uploadSoundSupport}
              languageCharactersProps={[
                { language: 'spanish' },
                { language: 'special' },
              ]}
              mathMlOptions={mathMlOptions}
            />
            {teacherInstructionsError && (
              <div className={classes.errorText}>{teacherInstructionsError}</div>
            )}
          </InputContainer>
        )}

        {promptEnabled && (
          <InputContainer
            label={prompt.label}
            className={classes.promptHolder}
          >
            <EditableHtml
              className={classes.prompt}
              markup={model.prompt}
              onChange={(value) => this.handleChange('prompt', value)}
              imageSupport={imageSupport}
              nonEmpty={false}
              disableUnderline
              error={promptError}
              toolbarOpts={toolbarOpts}
              pluginProps={getPluginProps(
                teacherInstructions?.inputConfiguration,
                baseInputConfiguration
              )}
              spellCheck={spellCheckEnabled}
              maxImageWidth={defaultImageMaxWidth}
              maxImageHeight={defaultImageMaxHeight}
              uploadSoundSupport={uploadSoundSupport}
              languageCharactersProps={[
                { language: 'spanish' },
                { language: 'special' },
              ]}
              mathMlOptions={mathMlOptions}
            />
            {promptError && (
              <div className={classes.errorText}>{promptError}</div>
            )}
          </InputContainer>
        )}

        <Typography className={classes.title}>Response Template</Typography>

        <EditableHtml
          activePlugins={ALL_PLUGINS}
          toolbarOpts={{ position: 'top' }}
          spellCheck={spellCheckEnabled}
          pluginProps={getPluginProps(
            template?.inputConfiguration,
            baseInputConfiguration
          )}
          responseAreaProps={{
            type: 'math-templated',
            respAreaToolbar: null,
            error: () => choicesErrors,
            onHandleAreaChange: this.onHandleAreaChange,
            maxResponseAreas: maxResponseAreas,
          }}
          className={classes.markup}
          markup={model.slateMarkup}
          onChange={this.onChange}
          imageSupport={imageSupport}
          disableImageAlignmentButtons={true}
          onBlur={this.onBlur}
          disabled={false}
          highlightShape={false}
          error={responseAreasError}
          uploadSoundSupport={uploadSoundSupport}
          languageCharactersProps={[
            { language: 'spanish' },
            { language: 'special' },
          ]}
          mathMlOptions={mathMlOptions}
        />

        <Typography className={classes.title}>Define Response</Typography>

        <InputContainer
          label="Response Template Equation Editor"
          className={classes.selectContainer}
        >
          <Select
            className={classes.select}
            onChange={(event) =>
              this.handleChange('equationEditor', event.target.value)
            }
            value={equationEditor}
          >
            <MenuItem value="non-negative-integers">
              Numeric - Non-Negative Integers
            </MenuItem>
            <MenuItem value="integers">Numeric - Integers</MenuItem>
            <MenuItem value="decimals">Numeric - Decimals</MenuItem>
            <MenuItem value="fractions">Numeric - Fractions</MenuItem>
            <MenuItem value={1}>Grade 1 - 2</MenuItem>
            <MenuItem value={3}>Grade 3 - 5</MenuItem>
            <MenuItem value={6}>Grade 6 - 7</MenuItem>
            <MenuItem value={8}>Grade 8 - HS</MenuItem>
            <MenuItem value={'geometry'}>Geometry</MenuItem>
            <MenuItem value={'advanced-algebra'}>
              Advanced Algebra
            </MenuItem>
            <MenuItem value={'statistics'}>Statistics</MenuItem>
            <MenuItem value={'item-authoring'}>Item Authoring</MenuItem>
          </Select>
        </InputContainer>

        {Object.keys(responses || {}).map((responseKey, idx) => {
          const response = responses[responseKey];

          return (
            <Response
              key={responseKey}
              responseKey={responseKey}
              mode={equationEditor}
              response={response}
              onResponseChange={this.onResponseChange}
              index={idx}
              cIgnoreOrder={cIgnoreOrder}
              cAllowTrailingZeros={cAllowTrailingZeros}
              error={responsesErrors && responsesErrors[idx]}
            />
          );
        })}

        {rationaleEnabled && (
          <InputContainer
            label={rationale.label}
            className={classes.promptHolder}
          >
            <EditableHtml
              className={classes.prompt}
              markup={model.rationale || ''}
              onChange={(value) => this.handleChange('rationale', value)}
              imageSupport={imageSupport}
              nonEmpty={false}
              error={rationaleError}
              toolbarOpts={toolbarOpts}
              pluginProps={getPluginProps(rationale?.inputConfiguration, {
                ...baseInputConfiguration,
                math: {
                  controlledKeypadMode: false,
                },
              })}
              spellCheck={spellCheckEnabled}
              maxImageWidth={
                (maxImageWidth && maxImageWidth.rationale) ||
                defaultImageMaxWidth
              }
              maxImageHeight={
                (maxImageHeight && maxImageHeight.rationale) ||
                defaultImageMaxHeight
              }
              uploadSoundSupport={uploadSoundSupport}
              languageCharactersProps={[
                { language: 'spanish' },
                { language: 'special' },
              ]}
              mathMlOptions={mathMlOptions}
            />
            {rationaleError && (
              <div className={classes.errorText}>{rationaleError}</div>
            )}
          </InputContainer>
        )}
      </layout.ConfigLayout>
    );
  }
}

export default withStyles((theme) => ({
  promptHolder: {
    width: '100%',
    paddingTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2,
  },
  prompt: {
    width: '100%',
  },
  errorText: {
    color: theme.palette.error.main,
    fontSize: '0.75rem',
    marginTop: theme.spacing.unit,
  },
  markup: {
    width: '100%',
    marginTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 3,
  },
  selectContainer: {
    width: '100%',
    marginTop: theme.spacing.unit * 2,
  },
  select: {
    width: '100%',
  },
  title: {
    fontSize: theme.typography.fontSize * 1.25,
    fontWeight: 'bold',
  }
}))(Design);
