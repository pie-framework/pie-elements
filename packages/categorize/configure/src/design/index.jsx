import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Typography from '@material-ui/core/Typography';
import {
  FeedbackConfig,
  InputContainer,
  layout,
  settings,
} from '@pie-lib/config-ui';
import {
  countInAnswer,
  ensureNoExtraChoicesInAnswer
} from '@pie-lib/categorize';
import EditableHtml from '@pie-lib/editable-html';
import { uid, withDragContext } from '@pie-lib/drag';
import debug from 'debug';

import Categories from './categories';
import AlternateResponses from './categories/alternateResponses';
import Choices from './choices';
import { Divider } from './buttons';
import {
  buildAlternateResponses,
  buildCategories
} from './builder';
import Header from './header';

const { Panel, toggle, radio } = settings;
const { Provider: IdProvider } = uid;

export class Design extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    model: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    uid: PropTypes.string,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired
    })
  };

  constructor(props) {
    super(props);
    this.uid = props.uid || uid.generateId();
  }

  updateModel = props => {
    const { model, onChange } = this.props;

    const updatedModel = {
      ...model,
      ...props
    };

    //Ensure that there are no extra choices in correctResponse, if the user has decided that only one choice may be used.
    updatedModel.correctResponse = ensureNoExtraChoicesInAnswer(
      updatedModel.correctResponse || [],
      updatedModel.choices
    );

    //clean categories
    updatedModel.categories = updatedModel.categories.map(c => ({
      id: c.id,
      label: c.label
    }));

    updatedModel.choices = updatedModel.choices.map(h => ({
      id: h.id,
      content: h.content,
      categoryCount: h.categoryCount
    }));

    onChange(updatedModel);
  };

  changeRationale = rationale => {
    const { model, onChange } = this.props;

    onChange({
      ...model,
      rationale
    });
  };

  changeTeacherInstructions = teacherInstructions => {
    const { model, onChange } = this.props;

    onChange({
      ...model,
      teacherInstructions
    });
  };

  changeFeedback = feedback => {
    this.updateModel({ feedback });
  };

  onAddAlternateResponse = () => {
    const { model: { correctResponse } } = this.props;

    this.updateModel({
      correctResponse: (correctResponse || []).map(cr => ({
        ...cr,
        alternateResponses: [
          ...(cr.alternateResponses || []),
          []
        ]
      }))
    });
  };

  onPromptChanged = prompt => this.updateModel({ prompt });

  onRemoveAlternateResponse = (index) => {
    const { model: { correctResponse } } = this.props;

    this.updateModel({
      correctResponse: (correctResponse || []).map((cr) => ({
        ...cr,
        alternateResponses: (cr.alternateResponses || []).filter((alt, altIndex) => altIndex !== index)
      }))
    });
  };

  countChoiceInCorrectResponse = choice => {
    const { model } = this.props;
    const out = countInAnswer(choice.id, model.correctResponse);
    return out;
  };

  render() {
    const { classes, className, model, imageSupport, configuration, onChange, onConfigurationChanged } = this.props;
    const {
      partialScoring = {},
      lockChoiceOrder = {},
      teacherInstructions = {},
      studentInstructions = {},
      rationale = {},
      scoringType = {},
      feedback = {},
      prompt = {}
    } = configuration || {};

    const config = model.config || {};
    config.choices = config.choices || { label: '', columns: 2 };

    const categories = buildCategories(
      model.categories || [],
      model.choices || [],
      model.correctResponse || []
    );

    const alternateResponses = buildAlternateResponses(
      model.categories || [],
      model.choices || [],
      model.correctResponse || []
    );

    const choices = model.choices.map(c => {
      c.correctResponseCount = this.countChoiceInCorrectResponse(c);
      return c;
    });

    return (
      <IdProvider value={this.uid}>
        <layout.ConfigLayout
          settings={
            <Panel
              model={model}
              onChangeModel={onChange}
              configuration={configuration}
              onChangeConfiguration={onConfigurationChanged}
              groups={{
                'Settings': {
                  partialScoring: partialScoring.settings &&
                    toggle(partialScoring.label),
                  lockChoiceOrder: lockChoiceOrder.settings &&
                  toggle(lockChoiceOrder.label),
                  'feedback.enabled': feedback.settings && toggle(feedback.label, true),
                },
                'Properties': {
                  'teacherInstructions.enabled': teacherInstructions.settings &&
                    toggle(teacherInstructions.label, true),
                  'studentInstructions.enabled': studentInstructions.settings &&
                    toggle(studentInstructions.label, true),
                  'rationale.enabled': rationale.settings &&
                    toggle(rationale.label, true),
                  scoringType: scoringType.settings &&
                    radio(scoringType.label, ['auto', 'rubric']),
                },
              }}
            />
          }
        >
          <div className={classNames(classes.design, className)}>

            {prompt.settings && (
              <InputContainer
                label={prompt.label}
                className={classes.promptHolder}
              >
                <EditableHtml
                  className={classes.prompt}
                  markup={model.prompt}
                  onChange={this.onPromptChanged}
                  imageSupport={imageSupport}
                  nonEmpty={!prompt.settings}
                  disableUnderline
                />
              </InputContainer>
            )}

            {teacherInstructions.enabled && (
              <InputContainer label={teacherInstructions.label} className={classes.inputHolder}>
                <EditableHtml
                  className={classes.input}
                  markup={model.teacherInstructions || ''}
                  onChange={this.changeTeacherInstructions}
                  imageSupport={imageSupport}
                  nonEmpty={false}
                />
              </InputContainer>
            )}

            {rationale.enabled && (
              <InputContainer label={rationale.label} className={classes.inputHolder}>
                <EditableHtml
                  className={classes.input}
                  markup={model.rationale || ''}
                  onChange={this.changeRationale}
                  imageSupport={imageSupport}
                  nonEmpty={false}
                />
              </InputContainer>
            )}

            <Categories
              imageSupport={imageSupport}
              model={model}
              categories={categories}
              onModelChanged={this.updateModel}
            />

            <Header
              className={classes.alternatesHeader}
              label="Alternate Responses"
              buttonLabel="ADD AN ALTERNATE RESPONSE"
              onAdd={this.onAddAlternateResponse}
            />

            {
              alternateResponses.map((categoriesList, index) => {
                return (
                  <React.Fragment key={index}>
                    <Header
                      className={classes.alternatesHeader}
                      label="Alternate Response"
                      buttonLabel="REMOVE ALTERNATE RESPONSE"
                      onAdd={() => this.onRemoveAlternateResponse(index)}
                    />

                    <AlternateResponses
                      altIndex={index}
                      imageSupport={imageSupport}
                      model={model}
                      categories={categoriesList}
                      onModelChanged={this.updateModel}
                    />

                  </React.Fragment>
                )
              })
            }

            <Divider />
            <Choices
              imageSupport={imageSupport}
              choices={choices}
              model={model}
              onModelChanged={this.updateModel}
            />

            {
              feedback.enabled && (
                <FeedbackConfig
                  feedback={model.feedback}
                  onChange={this.changeFeedback}
                />
              )
            }
          </div>
        </layout.ConfigLayout>
      </IdProvider>
    );
  }
}

const styles = theme => ({
  alternatesHeader: {
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2
  },
  text: {
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2
  },
  design: {
    paddingTop: theme.spacing.unit,
    paddingBottom: theme.spacing.unit
  },
  inputHolder: {
    width: '100%',
    paddingBottom: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2
  },
  input: {
    paddingTop: theme.spacing.unit * 2,
    width: '100%',
    maxWidth: '600px'
  },
  prompt: {
    paddingTop: theme.spacing.unit * 2,
    width: '100%'
  },
  promptHolder: {
    width: '100%',
    paddingBottom: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2
  },
  title: {
    marginBottom: '30px'
  }
});

export default withDragContext(withStyles(styles)(Design));
