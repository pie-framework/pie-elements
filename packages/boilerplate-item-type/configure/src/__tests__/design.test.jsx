import { shallow } from 'enzyme';
import React from 'react';
import { Design } from '../design';
import defaultValues from '../defaults';

jest.mock('@pie-lib/config-ui', () => ({
  layout: {
    ConfigLayout: (props) => <div>{props.children}</div>,
  },
  settings: {
    Panel: (props) => <div oChangeModel={props.onChange} />,
    toggle: jest.fn(),
    radio: jest.fn(),
  },
}));

jest.mock('lodash/debounce', () => (fn) => fn);

describe('design', () => {
  let w;
  let onChange;
  let onChangeConfig;

  const getModel = () => ({
    tokens: [],
  });
  beforeEach(() => {
    onChange = jest.fn();
    onChangeConfig = jest.fn();
    w = shallow(
      <Design
        model={getModel()}
        configuration={defaultValues.configuration}
        classes={{}}
        className={'foo'}
        onModelChanged={onChange}
        onConfigurationChanged={onChangeConfig}
      />,
    );
  });

  describe('snapshot', () => {
    it('renders all items with defaultProps', () => {
      expect(w).toMatchSnapshot();
    });

    it('tokenizer renders with html entities', () => {
      expect(
        shallow(
          <Design
            model={{
              text: '<p>&#8220;Lucy?&#63; Are you using your time wisely to plan your project?&#33;&#33;&#33;&#8221; Mr. Wilson asked.</p><p>Lucy looked a little confused at first. &#195; Then she grinned and proudly stated, &#8220;Why, yes I am! I plan to make a bird feeder for that tree out our window!&#8221;</p>',
              tokens: [],
            }}
            configuration={defaultValues.configuration}
            classes={{}}
            className={'foo'}
            onModelChanged={onChange}
            onConfigurationChanged={onChangeConfig}
          />,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    const assert = (fn, args, expected) => {
      const e = expected(getModel());

      it(`${fn} ${JSON.stringify(args)} => ${JSON.stringify(e)}`, () => {
        const instance = w.instance();
        instance[fn].apply(instance, args);

        expect(onChange).toBeCalledWith(e);
      });
    };

    describe('changePrompt', () => {
      assert('onPromptChanged', ['New Prompt'], (m) => ({
        ...m,
        prompt: 'New Prompt',
      }));
    });
  });
});
