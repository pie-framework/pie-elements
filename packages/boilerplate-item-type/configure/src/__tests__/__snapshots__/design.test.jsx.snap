// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design snapshot renders all items with defaultProps 1`] = `
<ConfigLayout
  dimensions={Object {}}
  settings={
    <Panel
      configuration={
        Object {
          "prompt": Object {
            "label": "Prompt",
            "settings": true,
          },
        }
      }
      groups={
        Object {
          "Properties": Object {
            "promptEnabled": undefined,
          },
          "Settings": Object {},
        }
      }
      model={
        Object {
          "tokens": Array [],
        }
      }
      onChangeConfiguration={[MockFunction]}
      onChangeModel={[Function]}
    />
  }
/>
`;

exports[`design snapshot tokenizer renders with html entities 1`] = `
<ConfigLayout
  dimensions={Object {}}
  settings={
    <Panel
      configuration={
        Object {
          "prompt": Object {
            "label": "Prompt",
            "settings": true,
          },
        }
      }
      groups={
        Object {
          "Properties": Object {
            "promptEnabled": undefined,
          },
          "Settings": Object {},
        }
      }
      model={
        Object {
          "text": "<p>&#8220;Lucy?&#63; Are you using your time wisely to plan your project?&#33;&#33;&#33;&#8221; Mr. Wilson asked.</p><p>Lucy looked a little confused at first. &#195; Then she grinned and proudly stated, &#8220;Why, yes I am! I plan to make a bird feeder for that tree out our window!&#8221;</p>",
          "tokens": Array [],
        }
      }
      onChangeConfiguration={[MockFunction]}
      onChangeModel={[Function]}
    />
  }
/>
`;
